File: Week6/ProblemSet6/sentimental-readability/readability.py
# This program computes the Coleman-Liau index as 0.0588 * L - 0.296 * S - 15.8
# where L is the average number of letters per 100 words in the text,
# and S is the average number of sentences per 100 words in the text.
# The output is the school grade level of difficulty to read a text given as input.
import re
from cs50 import get_string

# Get user input
ui = get_string("Text: ")

# Calculate Coleman-Liau index
words_count = ui.count(" ") + 1
sentences_count = ui.count(".") + ui.count("!") + ui.count("?")
letters_count = len(re.sub("[^A-Za-z]", "", ui))
S = 100 * sentences_count / words_count
L = 100 * letters_count / words_count
coleman_liau_index = 0.0588 * L - 0.296 * S - 15.8

# Determine grade based on index
grade = round(coleman_liau_index, 0)

if grade < 1:
    print("Before Grade 1")
elif grade <= 16:
    print(f"Grade {grade:.0f}")
else:
    print("Grade 16+")

File: Week2/ProblemSet2/readability/readability.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int count_letters(string L);
int count_sentences(string S);
int count_words(string text);

int main(void)
{
    string input = get_string("Text: ");
    int W = count_words(input); // W is the number of words
    double L = count_letters(input) * 100.00 /
               W; // not the average number of letters per 100 words in the text
    double S = count_sentences(input) * 100.00 /
               W; // not the average number of sentences per 100 words in the text

    int index = round(0.0588 * L - 0.296 * S - 15.8);

    // Printing Grade:
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", (int) index);
    }
}

// Additional Functions

int count_letters(string text)
{
    int L = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        char isupper = text[i] >= 'A' && text[i] <= 'Z';
        char islower = text[i] >= 'a' && text[i] <= 'z';
        if (isupper || islower)
        {
            L++;
        }
    }
    return L;
}

int count_sentences(string text)
{
    int S = 0;
    for (int i = 0; text[i]; i++)
    {
        char isdot = text[i] == '.' ? 1 : 0;
        char isquestion = text[i] == '?' ? 1 : 0;
        char isexclaimation = text[i] == '!' ? 1 : 0;
        if (isdot || isquestion || isexclaimation)
        {
            S++;
        }
    }
    return S;
}

int count_words(string text)
{
    int W = 1;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == ' ')
        {
            W++;
        }
    }

    if (text[0] == '\0')
    {
        W = 0;
    }
    return W;
}
